File: src/main/java/com/key/oa/config/CustomControllerAdvice.java
package com.key.oa.config;

import com.key.oa.common.BadCredentialsMessage;
import com.key.oa.common.JsonResponse;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.MalformedJwtException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

/**
 * 自定义全局异常处理机制
 *
 * @author 孙强
 */
@Slf4j
@RestControllerAdvice
public class CustomControllerAdvice {
    @ExceptionHandler(UsernameNotFoundException.class)
    public ResponseEntity<JsonResponse<Void>> usernameNotFountExceptionHandler(UsernameNotFoundException exception) {
        if (log.isInfoEnabled()) {
            log.info("Username not found: ", exception);
        }

        JsonResponse<Void> response = JsonResponse.error("A0201", "用户账户不存在");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
    }

    @ExceptionHandler(BadCredentialsException.class)
    public ResponseEntity<JsonResponse<Void>> badCredentialsExceptionHandler(BadCredentialsException exception) {
        JsonResponse<Void> response;

        if (BadCredentialsMessage.PASSWORD_WRONG.equals(exception.getMessage())) {
            if (log.isInfoEnabled()) {
                log.info("Password wrong.");
            }

            response = JsonResponse.error("A0210", "用户密码错误");
        } else if (BadCredentialsMessage.TOKEN_NOT_FOUND.equals(exception.getMessage())) {
            if (log.isInfoEnabled()) {
                log.info("Token not found.");
            }

            response = JsonResponse.error("A0301", "访问未授权");
        } else {
            if (log.isInfoEnabled()) {
                log.info("Request unauthorized.");
            }

            response = JsonResponse.error("A0300", "访问权限异常");
        }
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
    }

    @ExceptionHandler(MalformedJwtException.class)
    public ResponseEntity<JsonResponse<Void>> malformedJwtExceptionHandler(MalformedJwtException exception) {
        if (log.isInfoEnabled()) {
            log.info("Token is malformed: {}.", exception.getMessage());
        }

        JsonResponse<Void> response = JsonResponse.error("A0403", "Token非法");
        return ResponseEntity.status(HttpStatus.OK).body(response);
    }

    @ExceptionHandler(ExpiredJwtException.class)
    public ResponseEntity<JsonResponse<Void>> expiredJwtExceptionHandler(ExpiredJwtException exception) {
        if (log.isInfoEnabled()) {
            log.info("Token expired: {}.", exception.getMessage());
        }

        JsonResponse<Void> response = JsonResponse.error("A0404", "Token已过期");
        return ResponseEntity.status(HttpStatus.OK).body(response);
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<JsonResponse<Void>> illegalArgumentExceptionHandler(IllegalArgumentException exception) {
        if (log.isInfoEnabled()) {
            log.info("Request argument is illegal: {}.", exception.getMessage());
        }

        JsonResponse<Void> response = JsonResponse.error("A0402", "无效的用户输入");
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<JsonResponse<String>> globalExceptionHandler(Exception exception) {
        log.error("Exception occurred: ", exception);

        JsonResponse<String> response = JsonResponse.error("B0001", "系统执行出错", exception.getMessage());
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
    }
}

File: src/main/java/com/key/oa/config/CustomControllerAdvice.kt
package com.key.oa.config

import com.key.oa.common.JsonResponse
import com.key.oa.common.JsonResponse.Companion.error
import com.key.oa.common.PASSWORD_WRONG
import com.key.oa.common.TOKEN_NOT_FOUND
import com.key.oa.common.logger
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.MalformedJwtException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class CustomControllerAdvice {
    val log by logger()

    @ExceptionHandler(UsernameNotFoundException::class)
    fun usernameNotFound(exception: UsernameNotFoundException): ResponseEntity<JsonResponse<Unit>> {
        val response: JsonResponse<Unit> = error("A0201", "用户账户不存在")
        return ResponseEntity.ok().body(response)
    }

    @ExceptionHandler(BadCredentialsException::class)
    fun badCredentialsExceptionHandler(exception: BadCredentialsException): ResponseEntity<JsonResponse<Unit>> {
        val response: JsonResponse<Unit> = when (exception.message) {
            PASSWORD_WRONG -> error("A0210", "用户密码错误")
            TOKEN_NOT_FOUND -> error("A0301", "访问未授权")
            else -> {
                if (log.isInfoEnabled) {
                    log.info("Request unauthorized: {}.", exception.message, exception)
                }
                error("A0300", "访问权限异常")
            }
        }
        return ResponseEntity.ok().body(response)
    }

    @ExceptionHandler(MalformedJwtException::class)
    fun malformedJwtExceptionHandler(exception: MalformedJwtException): ResponseEntity<JsonResponse<Unit>> {
        val response: JsonResponse<Unit> = error("A0403", "Token非法")
        return ResponseEntity.ok().body(response)
    }

    @ExceptionHandler(ExpiredJwtException::class)
    fun expiredJwtExceptionHandler(exception: ExpiredJwtException): ResponseEntity<JsonResponse<Unit>> {
        val response: JsonResponse<Unit> = error("A0404", "Token已过期")
        return ResponseEntity.ok().body(response)
    }

    @ExceptionHandler(IllegalArgumentException::class)
    fun illegalArgumentExceptionHandler(exception: IllegalArgumentException): ResponseEntity<JsonResponse<Unit>> {
        if (log.isInfoEnabled) {
            log.info("Request argument is illegal: {}.", exception.message, exception)
        }
        val response: JsonResponse<Unit> = error("A0402", "无效的用户输入")
        return ResponseEntity.ok().body(response)
    }

    @ExceptionHandler(Exception::class)
    fun globalExceptionHandler(exception: Exception): ResponseEntity<JsonResponse<String>> {
        log.error("Exception occurred:\n ", exception)
        val response: JsonResponse<String> = error("B0001", "", exception.message ?: "")
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response)
    }
}